### DISCUSSION TITLE
R.app Mac GUI improvements and fixes

### DISCUSSION THREAD

**Discussion by s-u** [r-devel/r-project-sprint#12] [Ideas]
The R for Mac GUI is the most comprehensive platform-specific R GUI and has been pioneering a lot of features now present in other GUIs and commercial products, but in recent years it has not been updated to take advantage of any of the new macOS features. This idea is an umbrella for any work of the R for Mac GUI which includes:

- fixing any reported bugs related to the GUI
- improve modern features or visuals (e.g., better integration with the dark mode, modernize the setting panel)
- improve/add translations
- propose interesting features that can be implemented reasonably quickly (e.g. usablity/ergonomic features)

Please note that the work should be done on the [GitHub presence of the GUI](https://github.com/R-macos/Mac-GUI) since is it not directly tied to the R sources.

**Comment by hturner** [comment]
Thanks for these ideas @s-u. If I understand correctly, the first two ideas would require knowledge of Objective-C, which wasn't mentioned by any participant in their application. Among participants, I think Mac users may be in the minority and I'm not sure if there's anyone outside R Core with working knowledge of the Mac GUI. So even the fourth idea may not be viable for the sprint.

However, translations is definitely an area that could be worked on. Of the current languages with translations for the Mac GUI (de, it, fr, ja, nl), we have fluent speakers in de, fr and ja, plus we have speakers of other languages (notably es) that might be good to add. 

It would be fantastic if we could enable the translations to be made via the Weblate server (https://translate.rx.studio/projects/r-project/), which currently only supports the gettext translations in base R. There would be many advantages to this, most importantly that translators would not need to know anything about the underlying infrastructure and do not even need to be Mac users in order to contribute. I have checked and Weblate does support the strings files used in the Mac GUI: https://docs.weblate.org/en/latest/formats/apple.html#apple. There may be additional work required to generate the strings for Weblate and to then apply a patch back from Weblate, but presumably this is all possible with a bit of scripting.

So a potential project would have two components:
 - Set up Weblate to work with the strings files from the Mac GUI (collaborating with those working on #10)
 - Start testing out workflow with speakers of currently translated languages
 
 Stretch goals:
  - Add new Left-to-Right languages, e.g. Spanish
  - Support Right-to-Left languages, this would likely require some work on your side, see the [Apple docs](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/SupportingRight-To-LeftLanguages/SupportingRight-To-LeftLanguages.html#//apple_ref/doc/uid/10000171i-CH17-SW1). We have several Arabic speakers at the sprint that could contribute translations.
 
Would you be willing to participate in such a project? If so, would you be willing to make a PR for a proposal along these lines? (c.f. https://contributor.r-project.org/r-project-sprint-2023/projects/ideas-and-proposals.html#proposals).

**Comment by s-u** [comment]
Sure, I don't have any preference for tools that the translators will use, the translation mechanics is trivial, those are simply text files, the translators wouldn't have to re-generate the GUI necessarily. If we want to add languages, that's also fine, I can easily generate the templates for any languages we want to add.

It would be interesting to play with the RTL support. It may be easier than it sounds, since R code is still LTR so core UI elements are not flipped. Even identifying what changes in GUI layout (as opposed to just text) are required would be interesting.
